src/metadata_service/fam_metadata_service_direct.cpp

```diff
@@ -434,19 +434,24 @@ int Fam_Metadata_Service_Direct::Impl_::get_dataitem_KVS(
         diKVS *kvs = new diKVS();
         kvs->diNameKVS = dataitemNameKVS;
         kvs->diIdKVS = dataitemIdKVS;
-        pthread_mutex_lock(&kvsMapLock);
+        pthread_rwlock_init(&kvs->kvsLock, NULL);
+        pthread_rwlock_rdlock(&kvsMapLock);
         auto kvsObj = metadataKvsMap->find(regionId);
         if (kvsObj == metadataKvsMap->end()) {
+          pthread_rwlock_unlock(&kvsMapLock);
+          pthread_rwlock_wrlock(&kvsMapLock);
```

Comment:
It would be better to perform a find on the regionId after taking wrlock and before inserting the item in that kvsMap. Thought this code will have two times "find" but it will ensure synchronization between to threads trying to insert the item in the map.

You can do similar changes in other places, while inserting the item from the map.

 Delete case looks fine since you are using an additional lock on the kvs obj itself.